packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs

```diff
@@ -333,6 +424,429 @@ private MethodProvider BuildEmptyConstructor()
+            => Throw(New.Instance(
+                typeof(FormatException),
+                new FormattableStringExpression($"The model {{{0}}} does not support {(validationType == SerializationFormatValidationType.Write ? "writing" : "reading")} '{{{1}}}' format.",
+                [
+                    Nameof(typeOfT),
+                    format
+                ])));
+
+        /// <summary>
+        /// Constructs the body statements for the JsonModelWriteCore method containing the serialization for the model properties.
+        /// </summary>
+        private MethodBodyStatement[] CreateWritePropertiesStatements()
+        {
+            var propertyCount = _model.Properties.Count;
+            var propertyStatements = new MethodBodyStatement[propertyCount];
+            for (var i = 0; i < propertyCount; i++)
+            {
+                var property = _model.Properties[i];
+                var propertyMember = new MemberExpression(null, property.Name);
+                var propertySerializationInfo = property.SerializationInfo;
+                var propertySerializationFormat = propertySerializationInfo?.SerializationFormat ?? SerializationFormat.Default;
+                var propertyIsReadOnly = propertySerializationInfo?.IsReadOnly ?? false;
+                var propertyIsRequired = propertySerializationInfo?.IsRequired ?? false;
+
+                // Generate the serialization statements for the property
+                var writePropertySerializationStatements = new MethodBodyStatement[]
+                {
+                    _utf8JsonWriterSnippet.WritePropertyName(property.Name.ToVariableName()),
+                    CreateSerializationStatement(property.Type, propertyMember, propertySerializationFormat)
+                };
+
+                // Wrap the serialization statement in a check for whether the property is defined
+                var wrapInIsDefinedStatement = WrapInIsDefined(property, propertyMember, propertyIsRequired, writePropertySerializationStatements);
+                propertyStatements[i] = propertyIsReadOnly ?
+                    WrapInCheckNotWireIfStatement(_mrwOptionsParameterSnippet.Format, wrapInIsDefinedStatement)
+                    : wrapInIsDefinedStatement;
+            }
+
+            return propertyStatements;
+        }
+
+        /// <summary>
+        /// Wraps the serialization statement in a condition check to ensure only initialized and required properties are serialized.
+        /// </summary>
+        /// <param name="propertyProvider">The model property.</param>
+        /// <param name="propertyMemberExpression">The expression representing the property to serialize.</param>
+        /// <param name="writePropertySerializationStatement">The serialization statement to conditionally execute.</param>
+        /// <returns>A method body statement that includes condition checks before serialization.</returns>
+        private MethodBodyStatement WrapInIsDefined(
```

Comment:
nit: some of these methods do not appear to use instance methods / properties and can probably be converted to static for perf